flink-runtime/src/main/java/org/apache/flink/runtime/metrics/util/MetricUtils.java

```diff
@@ -214,6 +219,29 @@ static void instantiateNonHeapMemoryMetrics(final MetricGroup metricGroup) {
 		instantiateMemoryUsageMetrics(metricGroup, () -> ManagementFactory.getMemoryMXBean().getNonHeapMemoryUsage());
 	}
 
+	@VisibleForTesting
+	static void instantiateMetaspaceMemoryMetrics(final MetricGroup parentMetricGroup) {
+		final List<MemoryPoolMXBean> memoryPoolMXBeans = ManagementFactory.getMemoryPoolMXBeans()
+			.stream()
+			.filter(bean -> "Metaspace".equals(bean.getName()))
+			.collect(Collectors.toList());
+
+		if (memoryPoolMXBeans.isEmpty()) {
+			LOG.warn("No memory pool named 'Metaspace' is present. The '{}' metric group is not going to be instantiated.", METRIC_GROUP_METASPACE_NAME);
+			return;
+		}
+
+		final MetricGroup metricGroup = parentMetricGroup.addGroup(METRIC_GROUP_METASPACE_NAME);
+		final Iterator<MemoryPoolMXBean> beanIterator = memoryPoolMXBeans.iterator();
+
+		final MemoryPoolMXBean firstPool = beanIterator.next();
+		instantiateMemoryUsageMetrics(metricGroup, firstPool::getUsage);
+
+		if (beanIterator.hasNext()) {
+			LOG.warn("More than one memory pool named '{}' are present. Only the first pool was used for instantiating the metric.", METRIC_GROUP_METASPACE_NAME);
```

Comment:
I'm fairly sure that this is mostly a theoretical issue; for it to be usable via JMX the MBeans require some unique identification, and according to the MemoryPoolMXBean javadocs the only unique part of the identifier is in fact the name. In other words, multiple pools of this name being present would likely be a bug in the JVM.
We can keep this check, but it should be on debug imo.

It is also a bit odd to use the metric group name as the pool name, when we used another string for the actual filtering (as in, move "Metaspace" into a constant, use that here, and maybe make METRIC_GROUP_METASPACE_NAME an alias for this constant.)

Reply:
Yup, I bet it's quite theoretical. I just felt a bit uneasy about ignoring this one.

The pool name issue was actually not what I intended and, therefore, a "bug". Thanks for pointing that out.